src/v/storage/disk_log_impl.cc

```diff
@@ -1731,6 +1734,484 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+      .target = target,
+      .boundary = boundary,
+    };
+
+    storage::log_reader_config reader_cfg(index_entry.offset, target, priority);
+
+    reader_cfg.skip_batch_cache = true;
+    reader_cfg.skip_readers_cache = true;
+
+    auto reader = co_await make_reader(reader_cfg);
+    std::exception_ptr last_err;
+
+    try {
+        co_await std::move(reader).consume(acc, model::no_timeout);
+    } catch (...) {
+        last_err = std::current_exception();
+        vlog(stlog.error, "Error detected while consuming {}", last_err);
+    }
+    if (last_err) {
+        std::rethrow_exception(last_err);
+    }
+    co_return size_bytes;
+}
+
+ss::future<log::offset_range_size_result_t> disk_log_impl::offset_range_size(
+  model::offset first, model::offset last, ss::io_priority_class io_priority) {
+    vlog(
+      stlog.debug,
+      "Offset range size, first: {}, last: {}, lstat: {}",
+      first,
+      last,
+      offsets());
+    // build the collection
+    auto segments = [&] {
+        auto base_it = _segs.lower_bound(first);
+        std::vector<ss::lw_shared_ptr<storage::segment>> segments;
+        for (auto it = base_it; it != _segs.end(); it++) {
+            const auto& offsets = it->get()->offsets();
+            if (offsets.committed_offset < first) {
+                continue;
+            }
+            if (offsets.base_offset > last) {
+                break;
+            }
+            segments.push_back(*it);
+        }
+        return segments;
+    }();
+
+    // The following checks are needed to maintain the following invariants:
+    // - the 'first' offset of the offset range exists and the first segment in
+    // the
+    //   'segments' collection points to it.
+    // - if the log was truncated and start offset is greater than 'first' the
+    // method
+    //   should throw the exception.
+    // - the 'last' offset of the offset range exists and the the last segment
+    // in the
+    //   'segments' collection points to it.
+    // - if the last offset of the log is less than 'last' the method throws.
+    if (segments.empty()) {
+        vlog(stlog.error, "Can't find log segments to lock");
+        throw std::invalid_argument("Offset out of range");
+    }
+    if (segments.front()->offsets().base_offset > first) {
+        vlog(
+          stlog.error,
+          "Offset {} is out of range, {}",
+          first,
+          segments.front()->offsets());
+        throw std::invalid_argument("First offset is out of range");
+    }
+    if (segments.back()->offsets().committed_offset < last) {
+        vlog(
+          stlog.error,
+          "Offset {} is out of range, {}",
+          last,
+          segments.back()->offsets());
+        throw std::invalid_argument("Last offset is out of range");
+    }
+
+    std::vector<ss::future<ss::rwlock::holder>> f_locks;
+    f_locks.reserve(segments.size());
+    for (auto& s : segments) {
+        f_locks.emplace_back(s->read_lock());
```

Comment:
any chance of deadlocks here or we assume locks are always/everwhere taken in the increasing segment order?

Reply:
> yep

which one of @nvartolomei questions are you responding to?